### ISSUE TITLE
Use appropriate ADK DISM if available

### ISSUE THREAD

**Issue by brycem** [VirtualEngine/Lability#139]
On WS2012R2 with Windows ADK 8.1 and 10 installed, I'm getting the following error:

PS C:>  new-labvm -Name NANO -MediaId 2016TP5_x64_NANO_EN -SwitchName VmNAT
WARNING: [5:30:26 PM] Using existing 'VmNAT' virtual switch.
New-LabImage : To service this Windows image requires the latest version of the DISM. See
http://go.microsoft.com/fwlink/?LinkId=293395 to find the latest version of DISM,
and http://go.microsoft.com/fwlink/?LinkId=293394 to learn how to install the latest version of DISM from the ADK on
your computer.
At C:\Program Files\WindowsPowerShell\Modules\Lability\0.10.0\Src\LabVM.ps1:326 char:27
- ... ef] $null = New-LabImage -Id $node.Media -ConfigurationData $Configur ...
-                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  - CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
  - FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,New-LabImage

**Comment by iainbrighton** [comment]
It sounds like the ADK installs another DISM module. Can you run the following and post the output here?
```powershell
Get-Module DISM -ListAvailable | Format-List
Import-Module DISM -Force -PassThru | Format-List
```

**Comment by brycem** [comment]
Transcript attached - It seems that the problem is that path and DISM module need to be explicitly set\imported for it to work per these instructions:  https://technet.microsoft.com/en-us/library/hh825010.aspx...  The ADK seems to install the Win10 DISM bits in "side-by-side" mode rather than upgrading the system's default DISM.  The version numbers of the ADK and WS2K12R2 DISM modules are the same (ModuleVersion=3.0), but in the WS2K12R2 dism.psd1 the Verb-WindowsCapability commandlets are missing from the DISM module, and it relies on the path to find the dependent assemblies and resources so it won't use the ADK files.
[ws2k12r2-adk-dism.txt](https://github.com/VirtualEngine/Lability/files/471151/ws2k12r2-adk-dism.txt)
[ws2k12r2-default-dism.txt](https://github.com/VirtualEngine/Lability/files/471152/ws2k12r2-default-dism.txt)

**Comment by iainbrighton** [comment]
Hi @brycem I'm not sure why you're getting errors about the `CustomMedia.json` file. This should always be created in `$env:ALLUSERSPROFILE\Lability` and never the installation folder. However, are we saying that if the ADK is installed, we ___have___ to use the ADK version of the DISM module or the built-in one?

**Comment by brycem** [comment]
@iainbrighton - As I read the ADK documentations, I think you're supposed to use a Win8.1 DISM to manipulate Win8.1 images and a Win10 DISM to manipulate Win10 images, though in practice a Win10 DISM appears to be backward compatible.

**Comment by iainbrighton** [comment]
@brycem Right then. I have a 2012 R2 machine and have replicated the issue. I'll implement some changes/tests and then you know when they're ready to be taken for a spin 😉!

**Comment by iainbrighton** [comment]
OK - getting PowerShell to use the correct version of the DISM module turned out to be intricate than it should be. Here's it working on a 2012 R2 host with a Server 2016 Nano Server image 😃.
```powershell
PS C:\Users\Administrator> [environment]::OSVersion

Platform ServicePack Version    VersionString
-------- ----------- -------    -------------
 Win32NT             6.3.9600.0 Microsoft Windows NT 6.3.9600.0

PS C:\Users\Administrator> $lab = Import-Module Lability -Force -PassThru
PS C:\Users\Administrator> & $lab { Get-Module -Name Microsoft.Dism.Powershell | Select Path, Version }

Path                                                                                  Version
----                                                                                  -------
C:\Windows\System32\WindowsPowerShell\v1.0\Modules\Dism\Microsoft.Dism.PowerShell.dll 6.3.0.0

PS C:\Users\Administrator> Set-LabHostDefault -DismPath 'C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Deployment Tools\amd64\DISM' -Verbose
WARNING: [16:50:49] Updating the Dism version requires a session reload. Please restart the PowerShell session to
ensure the required Dism version is loaded.
VERBOSE: [16:50:49] Loaded Dism module version '6.3.0.0'.

ConfigurationPath       : C:\Lability\Configurations
DifferencingVhdPath     : C:\Lability\VMVirtualHardDisks
HotfixPath              : C:\Lability\Hotfixes
IsoPath                 : C:\Lability\ISOs
ModuleCachePath         : C:\ProgramData\Lability\Modules
ParentVhdPath           : C:\Lability\MasterVirtualHardDisks
ResourcePath            : C:\Lability\Resources
ResourceShareName       : Resources
DisableLocalFileCaching : False
EnableCallStackLogging  : False
DismPath                : C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Deployment
                          Tools\amd64\DISM\Microsoft.Dism.Powershell.dll

PS C:\Users\Administrator> & $lab { Get-Module -Name Microsoft.Dism.Powershell | Select Path, Version } | fl

Path    : C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Deployment
          Tools\amd64\DISM\Microsoft.Dism.Powershell.dll
Version : 6.3.0.0

PS C:\Users\Administrator> 
```
Correct DLL is referenced, but module isn't reloaded correctly. ___Restart of PowerShell session is REQUIRED after calling Set-LabHostDefault -DismPath.___
```powershell
PS C:\Users\Administrator> $lab = Import-Module Lability -Force -PassThru
PS C:\Users\Administrator> & $lab { Get-Module -Name Microsoft.Dism.Powershell | Select Path, Version } | fl

Path    : C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Deployment
          Tools\amd64\DISM\Microsoft.Dism.Powershell.dll
Version : 10.0.0.0

PS C:\Users\Administrator> New-LabImage -Id 2016_x64_Standard_Nano_EN_Eval -Force -Verbose

VERBOSE: [16:57:32] Removing disk image 'C:\Lability\MasterVirtualHardDisks\2016_x64_Standard_Nano_EN_Eval.vhdx'.
VERBOSE: [16:57:32] Resource 'C:\Lability\ISOs\2016_x64_EN_Eval.iso' checksum matches '18A4F00A675B0338F3C7C93C4F131BEB'.
VERBOSE: [16:57:32] Creating disk image 'Windows Server 2016 Standard Nano 64bit English Evaluation'.
VERBOSE: [16:57:32] Creating disk image 'C:\Lability\MasterVirtualHardDisks\2016_x64_Standard_Nano_EN_Eval.vhdx'.
VERBOSE: [16:57:33] Mounting disk image 'C:\Lability\MasterVirtualHardDisks\2016_x64_Standard_Nano_EN_Eval.vhdx'.
VERBOSE: [16:57:33] Initializing disk image 'C:\Lability\MasterVirtualHardDisks\2016_x64_Standard_Nano_EN_Eval.vhdx'.
VERBOSE: [16:57:34] Creating 'EFI' disk partition.
VERBOSE: [16:57:34] Formatting 'EFI' disk partition.
VERBOSE: [16:57:34] Creating 'MSR' disk partition.
VERBOSE: [16:57:34] Creating 'Windows' disk partition.
VERBOSE: [16:57:34] Formatting 'Windows' disk partition.
VERBOSE: [16:57:36] Mounting disk image 'C:\Lability\ISOs\2016_x64_EN_Eval.iso'.
VERBOSE: [16:57:36] Locating WIM image 'Windows Server 2016 SERVERSTANDARDNANO' index.
VERBOSE: [16:57:36] Applying Windows Image '1' to 'C:\Lability\MasterVirtualHardDisks\2016_x64_Standard_Nano_EN_Eval.vhdx'.
VERBOSE: [16:57:41] Adding Windows package(s) 'Microsoft-NanoServer-Guest-Package' to image 'E:\'.
VERBOSE: [16:57:41] Adding package 'Microsoft-NanoServer-Guest-Package' to image 'E'.
VERBOSE: [16:57:50] Adding localized Windows package(s) 'Microsoft-NanoServer-Guest-Package_en-US' to image 'E:\'.
VERBOSE: [16:57:50] Adding package 'Microsoft-NanoServer-Guest-Package_en-US' to image 'E'.
VERBOSE: [16:57:54] Adding Windows package(s) 'Microsoft-NanoServer-DSC-Package' to image 'E:\'.
VERBOSE: [16:57:54] Adding package 'Microsoft-NanoServer-DSC-Package' to image 'E'.
VERBOSE: [16:57:56] Adding localized Windows package(s) 'Microsoft-NanoServer-DSC-Package_en-US' to image 'E:\'.
VERBOSE: [16:57:56] Adding package 'Microsoft-NanoServer-DSC-Package_en-US' to image 'E'.
VERBOSE: [16:57:57] Dismounting disk image 'C:\Lability\ISOs\2016_x64_EN_Eval.iso'.
VERBOSE: [16:57:57] Repairing boot volume 'E:'.
VERBOSE: [16:57:57] Starting process 'bcdboot.exe E:\Windows /s D:\ /v /f UEFI'.
VERBOSE: [16:57:59] Process 'bcdboot.exe' exit code '0'.
PS C:\Users\Administrator>
```
Is there any chance you can check this [fork](https://github.com/iainbrighton/Lability/tree/Issue139) fixes your problems?

**Comment by tansei** [comment]
I've just stumble upon the same issue at the moment, and abslutelly puzzled at the moment.
Os w2k16 core hyper-v
```
Platform ServicePack    Version        VersionString
-------- -----------    -------        -------------
 Win32NT Service Pack 1 6.1.7601.65536 Microsoft Windows NT 6.1.7601 Service Pack 1
```
```
Path                                                                                  Version
----                                                                                  -------
C:\Windows\System32\WindowsPowerShell\v1.0\Modules\Dism\Microsoft.Dism.PowerShell.dll 10.0.0.0
```

```
DisableLocalFileCaching      : False
ResourcePath                 : C:\Lability\Resources
DifferencingVhdPath          : E:\Lability\VMVirtualHardDisks
EnableCallStackLogging       : False
ParentVhdPath                : E:\Lability\MasterVirtualHardDisks
IsoPath                      : C:\Lability\ISOs
ResourceShareName            : Resources
ConfigurationPath            : E:\Lability\Configurations
HotfixPath                   : E:\Lability\Hotfixes
ModuleCachePath              : C:\ProgramData\Lability\Modules
DismPath                     : C:\Windows\System32\WindowsPowerShell\v1.0\Modules\Dism\Microsoft.Dism.PowerShell.dll
RepositoryUri                : https://www.powershellgallery.com/api/v2/package
DisableSwitchEnvironmentName : True
```
and getting exact same warning once trying to provision w2k16 or w10 through lability :( 
any guideance / help / direction much appreciated. Thank you!

**Comment by iainbrighton** [comment]
@tansei What version of Lability are you using? There was an error creating Windows 10 images in 0.12.3 which should be fixed in 0.12.4.

**Comment by tansei** [comment]
@iainbrighton - thanks for your brilliant sessions in e2evc and channel9 videos, and thank you for the 0.12.4 indeed it is sorting out that problem and at least in my case the dism 10.0.0.0 does NOT appear anymore.
